proposals/metadata.md

```diff
@@ -0,0 +1,88 @@
+permissions and limitations under the License.
+-->
+
+### Container metadata service
+In order to solve the issues related to container metadata like: [#288](https://github.com/aws/amazon-ecs-agent/issues/288) and [#456](https://github.com/aws/amazon-ecs-agent/issues/456), we are considering adding a new feature "Container Metadata Service" similar to [EC2 Instance Metadata Service](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html) in ECS agent to provide the container metadata.
+
+By adding this feature, you can query the information about the task, container and container instance from within the container or from the container instance by reading the metadata file for each container. The following information will be available from this feature:
+ * Information from docker:
+  * Container NetworkSettings
+  * Container ID from docker
+  * Container Name
+  * Image Name
+  * Image ID
+  * Port Mapping
+ * Information from ECS:
+  * Cluster Arn
+  * Container Instance Arn
+  * Task Arn
+
+ECS agent will create a file for each container under the directory `/var/lib/ecs/data/` and then mount the file to the container under a directory eg: `/ecs/metadata` with read only permission when the container is being created. Also an environment variable denoting the file path eg: `ECS_CONTAINER_METADATA_FILE` will be injected into the container. After the container is started, ECS agent will update the metadata of container into this file. Then the container can get its metada information by parsing the json file with path `ECS_CONTAINER_METADATA_FILE`. This directory will be removed when task is being clean up. The detailed steps when start a container will be:
+* Create metadata file for each container under `/var/lib/ecs/data/` for each container.
+* Create the container with the two extra steps:
+  * Mount the metadata file created above to a specific directory eg: `/ecs/metadata/` with read only permission for each container.
+  * Inject the metadata file path as an environment variable to the container eg: `ECS_CONTAINER_METADATA_FILE=/ecs/metadata/container1`.
+* After the container is created, write the basic metadata information into the metadata file.
+* Start the container and update the networking information into the metadata file.
+* When the container is removed during task cleanup, remove the metadata file from disk.
+
+Information contained in the metadata file will look like:
+```json
+{
+  "Cluster": "test",
+  "Container_Instance": "arn:aws:ecs:us-west-2:xxxxxxxx:container-instance/732d1a0b-acbe-43e4-8358-f477d83eab03",
+  "TaskArn": "",
+  "ImageName": "ubuntu",
+  "ImageID": "04bbc221d46d3e5eb86c046f22878a8053247ce9b1e6b0a24495b171f9d20025",
+  "ContainerID": "ef3dabed9bef087ed147024a221d2d3a822e85d53c0ca4024d3d78e7913eeb64",
+  "PortMappings": [{
+    "ContainerPort" : 80,
+    "HostPort": 80,
+    "BindIP": "",
+    "Protocol": "tcp"
+   }],
+   "NetworkSettings": {
```

Comment:
I'm not sure that most of this `NetworkSettings` struct makes sense.  Much of the information here doesn't seem relevant (do people make logic changes based on whether `HairpinMode` is on?) and the most relevant parts are obtainable through standard Linux userspace utilities (finding out your own IP address and gateway, for example).  Probably the most useful thing here would be to figure out what network mode you're in; `bridge` (default on Linux), `nat` (default on Windows), `host`, or `eni` (as discussed in https://github.com/aws/amazon-ecs-agent/pull/701).